File: packages/ts-sdk/src/neuron-mcp.ts
/**
 * @cosmonapse/sdk — MCP-server Neuron
 *
 * Wrap **any stdio MCP server** as a Neuron. This is a thin client wrapper —
 * it does NOT implement an MCP server. It spawns an existing server as a
 * subprocess, speaks MCP over stdio via `@modelcontextprotocol/sdk`, and
 * exposes that server's tools behind the NeuronFn signature. A TASK becomes an
 * MCP `tools/call`; the tool result becomes the Neuron output.
 *
 * The `@modelcontextprotocol/sdk` package is an optional peer dependency and is
 * imported lazily, so projects that don't use MCP neurons don't need it.
 *
 * ```ts
 * import { Axon, mcpNeuron } from "@cosmonapse/sdk";
 *
 * const files = mcpNeuron({ server: "filesystem", args: ["/data"], tool: "read_file" });
 * const axon = new Axon({ neuronId: "files", neuronFn: files });
 * ```
 *
 * Input dict:
 *   tool              Tool to call (falls back to `opts.tool`, or the sole tool).
 *   arguments | args  Tool arguments (object). If omitted, all non-control keys
 *                     are used as arguments.
 *   __list_tools__    When truthy, return the server's tool catalogue.
 *
 * Output (`tools/call`):
 *   { response, result, is_error, content, meta:{tool,server,command} }
 * Output (`__list_tools__`):
 *   { tools: [{ name, description, input_schema }] }
 */

import type { Json } from "./envelope.js";
import type { CloseableNeuronFn } from "./neuron-express.js";

export interface McpNeuronOptions {
  /** Executable to spawn (e.g. "npx", "uvx"). Required unless `server` is set. */
  command?: string;
  /** Arguments. Appended after a preset's args when `server` is set. */
  args?: string[];
  /** Name of a standard server preset (see {@link standardMcpServers}). */
  server?: string;
  /** Extra environment variables for the subprocess. */
  env?: Record<string, string>;
  /** Working directory for the subprocess. */
  cwd?: string;
  /** Default tool name used when the input omits `tool`. */
  tool?: string;
  clientName?: string;
  clientVersion?: string;
}

/**
 * Launch specs for standard, separately-published MCP servers. We wrap them —
 * we do not ship them. Anything in `opts.args` is appended (e.g. allowed dirs
 * for filesystem, `--repository <path>` for git).
 */
export const standardMcpServers: Record<string, { command: string; args: string[]; note: string }> = {
  filesystem: {
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-filesystem"],
    note: "Append one or more allowed directories, e.g. args=['/data'].",
  },
  memory: {
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-memory"],
    note: "Knowledge-graph memory store.",
  },
  everything: {
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-everything"],
    note: "Reference server exercising every MCP feature; handy for tests.",
  },
  sequentialthinking: {
    command: "npx",
    args: ["-y", "@modelcontextprotocol/server-sequential-thinking"],
    note: "Structured step-by-step reasoning tool.",
  },
  fetch: {
    command: "uvx",
    args: ["mcp-server-fetch"],
    note: "Fetch a URL and return its content as markdown/text.",
  },
  git: {
    command: "uvx",
    args: ["mcp-server-git"],
    note: "Read/inspect a git repo. Append --repository <path>.",
  },
  time: {
    command: "uvx",
    args: ["mcp-server-time"],
    note: "Current time and timezone conversions.",
  },
};

const CONTROL_KEYS = new Set(["tool", "arguments", "args", "__list_tools__"]);

function resolveLaunch(opts: McpNeuronOptions): { command: string; args: string[] } {
  const extra = opts.args ?? [];
  if (opts.server != null) {
    const preset = standardMcpServers[opts.server];
    if (!preset) {
      const available = Object.keys(standardMcpServers).sort().join(", ");
      throw new Error(
        `Unknown MCP server preset '${opts.server}'. Available: ${available}. ` +
          `(Or pass command/args to wrap any other stdio MCP server.)`,
      );
    }
    return { command: opts.command ?? preset.command, args: [...preset.args, ...extra] };
  }
  if (opts.command == null) {
    throw new Error("mcpNeuron(...) needs either `command` (+optional `args`) or a `server` preset name.");
  }
  return { command: opts.command, args: extra };
}

interface McpClient {
  connect(transport: unknown): Promise<void>;
  listTools(): Promise<{ tools?: Array<{ name: string; description?: string; inputSchema?: unknown }> }>;
  callTool(req: { name: string; arguments: Record<string, unknown> }): Promise<{
    content?: Array<{ type?: string; text?: string }>;
    structuredContent?: unknown;
    isError?: boolean;
  }>;
  close(): Promise<void>;
}

export function mcpNeuron(opts: McpNeuronOptions): CloseableNeuronFn {
  const { command, args } = resolveLaunch(opts);
  let client: McpClient | null = null;
  let connecting: Promise<McpClient> | null = null;

  async function ensure(): Promise<McpClient> {
    if (client) return client;
    if (connecting) return connecting;
    connecting = (async () => {
      // Lazy, optional dependency — imported only when an MCP neuron runs.
      // Variable specifiers keep the import non-analyzable, so `tsc`/bundlers
      // don't require @modelcontextprotocol/sdk to be installed at build time.
      const clientSpec = "@modelcontextprotocol/sdk/client/index.js";
      const stdioSpec = "@modelcontextprotocol/sdk/client/stdio.js";
      const clientMod = (await import(clientSpec)) as Record<string, unknown>;
      const stdioMod = (await import(stdioSpec)) as Record<string, unknown>;
      const Client = (clientMod as { Client: new (info: unknown, opts: unknown) => McpClient }).Client;
      const StdioClientTransport = (stdioMod as {
        StdioClientTransport: new (params: unknown) => unknown;
      }).StdioClientTransport;

      const transport = new StdioClientTransport({
        command,
        args,
        ...(opts.env ? { env: opts.env } : {}),
        ...(opts.cwd ? { cwd: opts.cwd } : {}),
      });
      const c = new Client(
        { name: opts.clientName ?? "cosmonapse", version: opts.clientVersion ?? "0.2.0" },
        { capabilities: {} },
      );
      await c.connect(transport);
      client = c;
      return c;
    })();
    return connecting;
  }

  const fn = (async (input: Json, _context: unknown[]): Promise<Json> => {
    const c = await ensure();
    const inp = (input ?? {}) as Record<string, unknown>;

    if (inp.__list_tools__) {
      const res = await c.listTools();
      return {
        tools: (res.tools ?? []).map((t) => ({
          name: t.name,
          description: t.description ?? null,
          input_schema: (t.inputSchema as Json) ?? null,
        })),
      } as Json;
    }

    let tool = (inp.tool as string | undefined) ?? opts.tool;
    let toolArgs = (inp.arguments ?? inp.args) as Record<string, unknown> | undefined;
    if (toolArgs == null) {
      toolArgs = {};
      for (const [k, v] of Object.entries(inp)) {
        if (!CONTROL_KEYS.has(k)) toolArgs[k] = v;
      }
    }

    if (!tool) {
      const res = await c.listTools();
      const names = (res.tools ?? []).map((t) => t.name);
      if (names.length === 1) {
        tool = names[0];
      } else {
        throw new Error(
          `MCP Neuron could not determine which tool to call. Pass tool=... (server exposes: ${JSON.stringify(names)}).`,
        );
      }
    }

    const res = await c.callTool({ name: tool, arguments: toolArgs });
    const content = res.content ?? [];
    const texts = content.filter((x) => x?.text != null).map((x) => x.text as string);

    return {
      response: texts.join("\n"),
      result: (res.structuredContent as Json) ?? null,
      is_error: Boolean(res.isError),
      content: content as Json,
      meta: { tool, server: opts.server ?? null, command },
    } as Json;
  }) as CloseableNeuronFn;

  fn.close = async (): Promise<void> => {
    const c = client;
    client = null;
    connecting = null;
    if (c) {
      try {
        await c.close();
      } catch {
        /* teardown must not throw */
      }
    }
  };

  return fn;
}

File: packages/python-sdk/cosmonapse/_neuron_mcp.py
"""
cosmonapse._neuron_mcp
~~~~~~~~~~~~~~~~~~~~~~~
Wrap **any stdio MCP server** as a Neuron.

This does *not* implement an MCP server. It is a thin client wrapper: it spawns
an existing MCP server as a subprocess, speaks the Model Context Protocol over
stdio, and exposes that server's tools behind the ``NeuronFn`` signature. A
TASK becomes an MCP ``tools/call`` and the tool result becomes the Neuron's
output.

Usage
-----
Wrap a published server by command::

    from cosmonapse import Axon, Neuron

    fs = Neuron(
        source="mcp",
        command="npx",
        args=["-y", "@modelcontextprotocol/server-filesystem", "/data"],
        tool="read_file",                # optional default tool
    )
    axon = Axon(neuron_id="files", neuron_fn=fs)

…or use a preset for a standard server (see ``STANDARD_MCP_SERVERS``)::

    fs = Neuron(source="mcp", server="filesystem", args=["/data"])
    web = Neuron(source="mcp", server="fetch")
    git = Neuron(source="mcp", server="git", args=["--repository", "."])

The preset only supplies the launch ``command``/``args`` for a well-known,
separately-published server; anything you pass in ``args`` is appended.

Input dict
----------
================  ===========================================================
key               meaning
================  ===========================================================
``tool``          Tool name to call. Falls back to the ``tool=`` constructor
                  arg, or  -  if the server exposes exactly one tool  -  that tool.
``arguments`` / ``args``  Tool arguments (dict). If omitted, every input key
                  except ``tool``/``arguments``/``args`` is treated as an
                  argument.
``__list_tools__``  When truthy, return the server's tool catalogue instead of
                  calling a tool.
================  ===========================================================

Output dict
-----------
``tools/call``::

    {
        "response": "<concatenated text content>",
        "result": <structured content> | None,
        "is_error": False,
        "content": [ ... raw content blocks ... ],
        "meta": {"tool": ..., "server": ...},
    }

``__list_tools__``::

    {"tools": [{"name": ..., "description": ..., "input_schema": {...}}, ...]}

Soft dependency
---------------
The official ``mcp`` package is required (``pip install mcp``). It is not a
core dependency so projects that don't use MCP neurons don't pull it in.
"""

from __future__ import annotations

import asyncio
from typing import Any

from cosmonapse._neuron_base import _BaseNeuron

# Launch specs for standard, separately-published MCP servers. We wrap them  - 
# we do not ship them. Anything supplied in the constructor `args` is appended
# to the preset args (e.g. allowed directories for filesystem, repo for git).
STANDARD_MCP_SERVERS: dict[str, dict[str, Any]] = {
    # Node-based reference servers (run via npx).
    "filesystem": {
        "command": "npx",
        "args": ["-y", "@modelcontextprotocol/server-filesystem"],
        "note": "Append one or more allowed directories, e.g. args=['/data'].",
    },
    "memory": {
        "command": "npx",
        "args": ["-y", "@modelcontextprotocol/server-memory"],
        "note": "Knowledge-graph memory store.",
    },
    "everything": {
        "command": "npx",
        "args": ["-y", "@modelcontextprotocol/server-everything"],
        "note": "Reference server exercising every MCP feature; handy for tests.",
    },
    "sequentialthinking": {
        "command": "npx",
        "args": ["-y", "@modelcontextprotocol/server-sequential-thinking"],
        "note": "Structured step-by-step reasoning tool.",
    },
    # Python-based reference servers (run via uvx).
    "fetch": {
        "command": "uvx",
        "args": ["mcp-server-fetch"],
        "note": "Fetch a URL and return its content as markdown/text.",
    },
    "git": {
        "command": "uvx",
        "args": ["mcp-server-git"],
        "note": "Read/inspect a git repo. Append --repository <path>.",
    },
    "time": {
        "command": "uvx",
        "args": ["mcp-server-time"],
        "note": "Current time and timezone conversions.",
    },
}

_CONTROL_KEYS = {"tool", "arguments", "args", "__list_tools__"}


def _require_mcp():
    try:
        import mcp  # noqa: F401
        return mcp
    except ImportError:
        raise ImportError(
            "The 'mcp' package is required for MCP-server Neuron wrappers.\n"
            "Install it with:  pip install mcp"
        ) from None


def _resolve_launch(
    command: str | None,
    args: list[str] | None,
    server: str | None,
) -> tuple[str, list[str]]:
    """Combine an explicit command / a named preset / extra args into a spec."""
    extra = list(args or [])

    if server is not None:
        preset = STANDARD_MCP_SERVERS.get(server)
        if preset is None:
            available = ", ".join(sorted(STANDARD_MCP_SERVERS))
            raise ValueError(
                f"Unknown MCP server preset {server!r}. Available: {available}. "
                f"(Or pass command=/args= to wrap any other stdio MCP server.)"
            )
        base_cmd = command or preset["command"]
        return base_cmd, [*preset["args"], *extra]

    if command is None:
        raise ValueError(
            "Neuron(source='mcp', ...) needs either `command` (+optional `args`) "
            "or a `server` preset name."
        )
    return command, extra


def _extract_call_result(result: Any) -> dict[str, Any]:
    """Normalise an MCP CallToolResult into a plain Neuron-output dict."""
    texts: list[str] = []
    content = getattr(result, "content", None) or []
    for item in content:
        text = getattr(item, "text", None)
        if text is not None:
            texts.append(text)

    structured = getattr(result, "structuredContent", None)
    is_error = bool(getattr(result, "isError", False))

    try:
        raw_content = [
            c.model_dump(mode="json") if hasattr(c, "model_dump") else c
            for c in content
        ]
    except Exception:
        raw_content = []

    return {
        "response": "\n".join(texts),
        "result": structured,
        "is_error": is_error,
        "content": raw_content,
    }


def _serialise_tools(result: Any) -> dict[str, Any]:
    tools = getattr(result, "tools", None) or []
    out = []
    for t in tools:
        out.append(
            {
                "name": getattr(t, "name", None),
                "description": getattr(t, "description", None),
                "input_schema": getattr(t, "inputSchema", None),
            }
        )
    return {"tools": out}


class _MCPNeuron(_BaseNeuron):
    """Expose a stdio MCP server's tools as a Neuron.

    A single long-lived background task owns the stdio connection and the MCP
    ``ClientSession``; requests are funnelled to it through a queue. This keeps
    every anyio cancel scope inside one task (the MCP stdio client requires the
    connection to be opened and closed from the same task), while letting many
    TASKs reuse the one spawned subprocess.
    """

    def __init__(
        self,
        command: str | None = None,
        args: list[str] | None = None,
        server: str | None = None,
        env: dict[str, str] | None = None,
        cwd: str | None = None,
        tool: str | None = None,
        client_name: str = "cosmonapse",
        client_version: str = "0.2.0",
    ) -> None:
        _require_mcp()  # fail fast with a clear message
        self.command, self.args = _resolve_launch(command, args, server)
        self.server = server
        self.env = env
        self.cwd = cwd
        self.tool = tool
        self.client_name = client_name
        self.client_version = client_version

        self._req_q: asyncio.Queue[Any] | None = None
        self._worker: asyncio.Task[Any] | None = None
        self._start_lock = asyncio.Lock()

    # ------------------------------------------------------------------
    # Public NeuronFn entry point
    # ------------------------------------------------------------------

    async def __call__(self, input: dict[str, Any], context: list[Any]) -> dict[str, Any]:
        if input.get("__list_tools__"):
            result = await self._request("list", {})
            return _serialise_tools(result)

        tool = input.get("tool") or self.tool
        arguments = input.get("arguments")
        if arguments is None:
            arguments = input.get("args")
        if arguments is None:
            arguments = {k: v for k, v in input.items() if k not in _CONTROL_KEYS}

        if not tool:
            # No explicit tool: if the server exposes exactly one, use it.
            listing = _serialise_tools(await self._request("list", {}))
            names = [t["name"] for t in listing["tools"]]
            if len(names) == 1:
                tool = names[0]
            else:
                raise ValueError(
                    "MCP Neuron could not determine which tool to call. "
                    f"Pass tool=... (server exposes: {names})."
                )

        result = await self._request("call", {"tool": tool, "arguments": arguments})
        out = _extract_call_result(result)
        out["meta"] = {"tool": tool, "server": self.server, "command": self.command}
        return out

    # ------------------------------------------------------------------
    # Background session runner
    # ------------------------------------------------------------------

    async def _ensure_started(self) -> None:
        if self._worker is not None and not self._worker.done():
            return
        async with self._start_lock:
            if self._worker is not None and not self._worker.done():
                return
            loop = asyncio.get_event_loop()
            self._req_q = asyncio.Queue()
            started: asyncio.Future[None] = loop.create_future()
            self._worker = asyncio.create_task(self._run(started))
            await started  # propagates startup errors (e.g. command not found)

    async def _run(self, started: asyncio.Future[None]) -> None:
        from mcp import ClientSession, StdioServerParameters
        from mcp.client.stdio import stdio_client

        params = StdioServerParameters(
            command=self.command,
            args=self.args,
            env=self.env,
            cwd=self.cwd,
        )
        try:
            async with stdio_client(params) as (read, write):
                async with ClientSession(read, write) as session:
                    await session.initialize()
                    if not started.done():
                        started.set_result(None)
                    assert self._req_q is not None
                    while True:
                        item = await self._req_q.get()
                        if item is None:  # shutdown sentinel
                            break
                        fut, kind, payload = item
                        try:
                            if kind == "list":
                                res = await session.list_tools()
                            else:
                                res = await session.call_tool(
                                    payload["tool"], payload["arguments"]
                                )
                            if not fut.done():
                                fut.set_result(res)
                        except Exception as exc:  # noqa: BLE001
                            if not fut.done():
                                fut.set_exception(exc)
        except Exception as exc:  # noqa: BLE001  -  startup / transport failure
            if not started.done():
                started.set_exception(exc)
            else:
                # Fail any in-flight request so callers don't hang forever.
                self._drain_with_error(exc)

    def _drain_with_error(self, exc: BaseException) -> None:
        if self._req_q is None:
            return
        while not self._req_q.empty():
            item = self._req_q.get_nowait()
            if item is None:
                continue
            fut = item[0]
            if not fut.done():
                fut.set_exception(exc)

    async def _request(self, kind: str, payload: dict[str, Any]) -> Any:
        await self._ensure_started()
        assert self._req_q is not None
        loop = asyncio.get_event_loop()
        fut: asyncio.Future[Any] = loop.create_future()
        await self._req_q.put((fut, kind, payload))
        return await fut

    # ------------------------------------------------------------------
    # Teardown
    # ------------------------------------------------------------------

    async def aclose(self) -> None:
        worker = self._worker
        if worker is None:
            return
        if self._req_q is not None and not worker.done():
            await self._req_q.put(None)
        try:
            await asyncio.wait_for(asyncio.shield(worker), timeout=10)
        except (asyncio.TimeoutError, asyncio.CancelledError):
            worker.cancel()
        except Exception:  # noqa: BLE001
            pass
        finally:
            self._worker = None
            self._req_q = None
